Create indexes in django models and migration files

acwhite211 opened the pull request:
Fixes #7482

Create indexes mentioned in the issue into the models and migration files.  Some of the indexes mentioned were already present in the existing model, the rest have been added.

The main downside I could see is with adding the tree field indexes is that writes might be too slow.  With an index like `taxon.name`, write operations like INSERT, UPDATE, and DELETE will take longer, but with the upside of read operations being faster.  I know that some of our tree operations make bulk edits to the tree record fields, like the 'Move' action in the tree viewer, so we'll want to be careful in our performance evaluation testing.  We'll want to test this on large databases with a big taxon tree to make sure the read and write performance is acceptable.

I ran into a problem with the tree viewer timing out after running the index migrations.  Solved the issue by rewriting the `get_tree_rows()` function to avoid the expensive grouped self-join on tree tables. Instead of joining child and synonym rows and collapsing them with GROUP BY, it now computes child counts and synonym lists with correlated subqueries, which preserves the same response shape while producing a faster query for taxon tree requests.

## Indexed fields

| Table | Indexed field(s) |
| --- | --- |
| `agentidentifier` | `identifier`, `identifiertype` |
| `agentspecialty` | `ordernumber`, `specialtyname` |
| `agentvariant` | `name` |
| `attachmentmetadata` | `name` |
| `author` | `ordernumber` |
| `collectionobject` | `name`, `projectnumber` |
| `collectionobjectgroup` | `guid`, `name` |
| `collectionobjectgrouptype` | `name` |
| `collectionobjectproperty` | `guid` |
| `collectionobjecttype` | `name` |
| `collectionreltype` | `name` |
| `exchangein` | `exchangeinnumber` |
| `exsiccataitem` | `number` |
| `geography` | `commonname`, `guid`, `highestchildnodenumber`, `nodenumber` |
| `geographytreedef` | `name` |
| `geographytreedefitem` | `name` |
| `geologictimeperiod` | `highestchildnodenumber`, `nodenumber` |
| `geologictimeperiodtreedef` | `name` |
| `geologictimeperiodtreedefitem` | `name` |
| `institutionnetwork` | `altname` |
| `latlonpolygon` | `name` |
| `lithostrat` | `highestchildnodenumber`, `nodenumber` |
| `lithostrattreedef` | `name` |
| `lithostrattreedefitem` | `name` |
| `locality` | `guid` |
| `materialsample` | `guid` |
| `morphbankview` | `viewname` |
| `otheridentifier` | `identifier` |
| `picklist` | `fieldname`, `filterfieldname`, `tablename` |
| `preparationproperty` | `guid` |
| `preptype` | `name` |
| `referencework` | `librarynumber` |
| `spauditlogfield` | `fieldname` |
| `specifyuser` | `name` |
| `spexportschema` | `schemaname` |
| `spexportschemaitem` | `fieldname` |
| `spexportschemaitemmapping` | `exportedfieldname` |
| `spexportschemamapping` | `mappingname` |
| `spfieldvaluedefault` | `fieldname`, `tablename` |
| `splocalecontainer` | `picklistname` |
| `splocalecontaineritem` | `picklistname`, `weblinkname` |
| `sppermission` | `name` |
| `spprincipal` | `name` |
| `spquery` | `contextname` |
| `spqueryfield` | `fieldname`, `formatname` |
| `spviewsetobj` | `filename` |
| `storage` | `highestchildnodenumber`, `nodenumber` |
| `storagetreedef` | `name` |
| `storagetreedefitem` | `name` |
| `taxon` | `cultivarname`, `groupnumber`, `highestchildnodenumber`, `nodenumber` |
| `taxontreedef` | `name` |
| `taxontreedefitem` | `name` |
| `tectonicunit` | `fullname`, `guid`, `highestchildnodenumber`, `name`, `nodenumber` |
| `tectonicunittreedef` | `name` |
| `tectonicunittreedefitem` | `name` |
| `voucherrelationship` | `vouchernumber` |

### Checklist

- [ ] Self-review the PR after opening it to make sure the changes look good and
      self-explanatory (or properly documented)
- [ ] Add relevant issue to release milestone
- [ ] Add pr to documentation list
- [ ] Add automated tests
- [ ] Add a reverse migration if a migration is present in the PR

### Testing instructions

- [ ] See that the new migration step for adding the indexes completes successfully.
- [ ] Use the QB on fields that have been indexed to see that they run correctly and in a timely manner.
- [ ] Use the taxon tree viewer on a database with a large taxon tree.  Try all of the tree operations to see that they run in a timely manner.

acwhite211 committed:
create indexes in models and migration files

acwhite211 committed:
Merge branch 'main' into issue-7482

acwhite211 committed:
Lint code with ESLint and Prettier

Triggered by 66a7b8420f26da7ac42ca751853ee1194c1288ce on branch refs/heads/issue-7482

acwhite211 committed:
update migration name

acwhite211 committed:
Fix field naming and remove indexes that aren't needed

acwhite211 committed:
Fix index name length errors

acwhite211 committed:
Rewrite get_tree_rows to avoid api timeout

acwhite211 committed:
Fix subqueries in get_tree_rows

acwhite211 committed:
Add test_taxon_rows_include_author_and_synonyms unit test

alesan99 reviewed (approved):
- [x] See that the new migration step for adding the indexes completes successfully.
- [x] Use the QB on fields that have been indexed to see that they run correctly and in a timely manner.
- [x] Use the taxon tree viewer on a database with a large taxon tree.  Try all of the tree operations to see that they run in a timely manner.
---
Migration runs correctly on KUBirds and read operations are snappy 👍
Tried deleting, moves, merges, searching, and importing big trees.
I didn't notice any speed drops either when running this locally.

acwhite211 committed:
Merge branch 'main' into issue-7482

acwhite211 committed:
Lint code with ESLint and Prettier

Triggered by f883c1c5e99af2729d49c280d531ed1924632d28 on branch refs/heads/issue-7482

bhumikaguptaa reviewed (commented):
- [ ] See that the new migration step for adding the indexes completes successfully.
- [ ] Use the QB on fields that have been indexed to see that they run correctly and in a timely manner.
- [X] Use the taxon tree viewer on a database with a large taxon tree.  Try all of the tree operations to see that they run in a timely manner.

---
When I tried to run the MaterialSample table I got the following error:

[Specify 7 Crash Report - 2026-03-18T18_28_32.334Z.txt](https://github.com/user-attachments/files/26096121/Specify.7.Crash.Report.-.2026-03-18T18_28_32.334Z.txt)

Link to DB: https://ojsmnh20251211-issue-7482.test.specifysystems.org/specify/query

acwhite211 committed:
handle MultipleObjectsReturned

bhumikaguptaa reviewed (approved):
- [X] See that the new migration step for adding the indexes completes successfully.
- [X] Use the QB on fields that have been indexed to see that they run correctly and in a timely manner.
- [X] Use the taxon tree viewer on a database with a large taxon tree.  Try all of the tree operations to see that they run in a timely manner.

--

It works as expected. I was able to query on indexed fields without any errors, including materialsample.

grantfitzsimmons committed:
Merge branch 'main' into issue-7482

grantfitzsimmons committed:
Lint code with ESLint and Prettier

Triggered by 0f0f17afa60a6bf4dab47d5600a0fca8495eaaef on branch refs/heads/issue-7482

acwhite211 commented:
Working on doing some more thorough testing to make sure that no action or process in Specify is significantly slowed down from re-indexing.

acwhite211 committed:
Merge branch 'main' into issue-7482

grantfitzsimmons reviewed (approved):
- [x] See that the new migration step for adding the indexes completes successfully.
- [x] Use the QB on fields that have been indexed to see that they run correctly and in a timely manner.
- [x] Use the taxon tree viewer on a database with a large taxon tree.  Try all of the tree operations to see that they run in a timely manner.
---

I did the same as @alesan99–  made some really big tree moves and imports. I don't feel the speed difference, but waiting for your testing to be done @acwhite211

grantfitzsimmons committed:
Merge branch 'main' into issue-7482

grantfitzsimmons committed:
Merge branch 'main' into issue-7482
